File: crates/braat-dsp/src/burg_lpc.rs
/* Braat, adapted from Praat
 * Copyright (C) 2026 Jocelyn Stericker <jocelyn@nettek.ca>
 * Copyright (C) 1992-2008,2010-2012,2014-2021,2024-2026 Paul Boersma
 * Copyright (C) 1993-2020 David Weenink
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 */

/// Burg's method for LPC coefficient estimation.
/// Source: N. Anderson in Childers (ed.), Modern Spectrum Analysis, IEEE Press, 1978.
/// Returns residual mean-square power (0 if ill-conditioned).
pub fn burg_lpc(
    frame: &[f32],
    frame_len: usize,
    order: usize,
    coeffs: &mut [f32],
    b1: &mut [f32],
    b2: &mut [f32],
    aa: &mut [f32],
) -> f32 {
    for j in 0..order {
        coeffs[j] = 0.0;
    }
    if frame_len <= 2 {
        coeffs[0] = -1.0;
        return if frame_len == 2 {
            0.5 * (frame[0] * frame[0] + frame[1] * frame[1])
        } else {
            frame[0] * frame[0]
        };
    }

    let mut p = 0.0f32;
    for j in 0..frame_len {
        p += frame[j] * frame[j];
    }
    let mut xms = p / frame_len as f32;
    if xms <= 0.0 {
        return 0.0;
    }

    // Initialise forward/backward prediction error vectors
    b1[0] = frame[0];
    for j in 1..frame_len - 1 {
        b1[j] = frame[j];
        b2[j - 1] = frame[j];
    }
    b2[frame_len - 2] = frame[frame_len - 1];

    for ii in 0..order {
        let active = frame_len - ii - 1;
        let mut num = 0.0f32;
        let mut denom = 0.0f32;
        for j in 0..active {
            num += b1[j] * b2[j];
            denom += b1[j] * b1[j] + b2[j] * b2[j];
        }
        if denom <= 0.0 {
            return 0.0;
        }
        coeffs[ii] = 2.0 * num / denom;

        xms *= 1.0 - coeffs[ii] * coeffs[ii];

        // Levinson recursion
        for j in 0..ii {
            coeffs[j] = aa[j] - coeffs[ii] * aa[ii - 1 - j];
        }

        if ii < order - 1 {
            for j in 0..=ii {
                aa[j] = coeffs[j];
            }
            let rc = aa[ii];
            let bound = active - 1;
            for j in 0..bound {
                b1[j] -= rc * b2[j];
                b2[j] = b2[j + 1] - rc * b1[j + 1];
            }
        }
    }
    xms
}

/// Evaluates the LPC polynomial P(z) = z^n - c[0]*z^(n-1) - ... - c[n-1]
/// and its derivative at complex z, writing [P.re, P.im, P'.re, P'.im] into out.
fn eval_lpc_poly(
    z_re: f32,
    z_im: f32,
    lpc_coeffs: &[f32],
    order: usize,
    out: &mut [f32; 4],
) {
    let mut p_re = 1.0f32;
    let mut p_im = 0.0f32;
    let mut dp_re = 0.0f32;
    let mut dp_im = 0.0f32;
    for k in 0..order {
        let ndp_re = dp_re * z_re - dp_im * z_im + p_re;
        let ndp_im = dp_re * z_im + dp_im * z_re + p_im;
        dp_re = ndp_re;
        dp_im = ndp_im;
        let np_re = p_re * z_re - p_im * z_im - lpc_coeffs[k];
        let np_im = p_re * z_im + p_im * z_re;
        p_re = np_re;
        p_im = np_im;
    }
    out[0] = p_re;
    out[1] = p_im;
    out[2] = dp_re;
    out[3] = dp_im;
}

/// Durand-Kerner root finding + Newton-Raphson polish.
/// Returns true if converged.
pub fn find_lpc_roots(
    lpc_coeffs: &[f32],
    order: usize,
    roots_re: &mut [f32],
    roots_im: &mut [f32],
    peval: &mut [f32; 4],
) -> bool {
    use std::f32::consts::PI;
    for k in 0..order {
        let angle = (2.0 * PI * (k as f32 + 0.5)) / order as f32;
        roots_re[k] = 0.9 * angle.cos();
        roots_im[k] = 0.9 * angle.sin();
    }

    let mut converged = false;
    'outer: for _iter in 0..80 {
        let mut max_step2 = 0.0f32;
        for k in 0..order {
            eval_lpc_poly(roots_re[k], roots_im[k], lpc_coeffs, order, peval);
            let mut den_re = 1.0f32;
            let mut den_im = 0.0f32;
            for j in 0..order {
                if j == k {
                    continue;
                }
                let d_re = roots_re[k] - roots_re[j];
                let d_im = roots_im[k] - roots_im[j];
                let n_re = den_re * d_re - den_im * d_im;
                den_im = den_re * d_im + den_im * d_re;
                den_re = n_re;
            }
            let d2 = den_re * den_re + den_im * den_im;
            if d2 == 0.0 {
                continue;
            }
            let step_re = (peval[0] * den_re + peval[1] * den_im) / d2;
            let step_im = (peval[1] * den_re - peval[0] * den_im) / d2;
            roots_re[k] -= step_re;
            roots_im[k] -= step_im;
            let s2 = step_re * step_re + step_im * step_im;
            if s2 > max_step2 {
                max_step2 = s2;
            }
        }
        if max_step2 < 1e-20 {
            converged = true;
            break 'outer;
        }
    }

    // Newton-Raphson polish
    for k in 0..order {
        let mut best_re = roots_re[k];
        let mut best_im = roots_im[k];
        eval_lpc_poly(best_re, best_im, lpc_coeffs, order, peval);
        let mut ymin = (peval[0] * peval[0] + peval[1] * peval[1]).sqrt();

        for _iter in 0..80 {
            eval_lpc_poly(roots_re[k], roots_im[k], lpc_coeffs, order, peval);
            let fabsy = (peval[0] * peval[0] + peval[1] * peval[1]).sqrt();
            if fabsy >= ymin {
                roots_re[k] = best_re;
                roots_im[k] = best_im;
                break;
            }
            ymin = fabsy;
            best_re = roots_re[k];
            best_im = roots_im[k];
            let dp2 = peval[2] * peval[2] + peval[3] * peval[3];
            if dp2 == 0.0 {
                break;
            }
            roots_re[k] -= (peval[0] * peval[2] + peval[1] * peval[3]) / dp2;
            roots_im[k] -= (peval[1] * peval[2] - peval[0] * peval[3]) / dp2;
        }
    }
    converged
}

/// Projects roots outside the unit circle to their inverse.
pub fn fix_into_unit_circle(roots_re: &mut [f32], roots_im: &mut [f32], n: usize) {
    for i in 0..n {
        let n2 = roots_re[i] * roots_re[i] + roots_im[i] * roots_im[i];
        if n2 > 1.0 {
            roots_re[i] /= n2;
            roots_im[i] = -roots_im[i] / n2;
        }
    }
}

File: src/lib/dsp/burgLpc.ts
/* Braat, adapted from Praat
 * Copyright (C) 2026 Jocelyn Stericker <jocelyn@nettek.ca>
 * Copyright (C) 1992-2008,2010-2012,2014-2021,2024-2026 Paul Boersma
 * Copyright (C) 1993-2020 David Weenink
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * Burg's method for autoregressive (LPC) coefficient estimation.
 *
 * Source: N. Anderson in Childers (ed.), Modern Spectrum Analysis,
 * IEEE Press, 1978, pp. 252-255.  Ported from dwsys/NUM2.cpp VECburg().
 *
 * The resulting analysis polynomial is:
 *   P(z) = z^order − coeffs[0]·z^(order−1) − coeffs[1]·z^(order−2) − ⋯ − coeffs[order−1]
 * Its roots are the LPC poles; formant frequencies come from poles with positive
 * imaginary part and angle in [safetyHz, nyquist − safetyHz].
 *
 * @param frame    Windowed, pre-emphasised mono samples, 0-indexed, length `frameLen`.
 * @param frameLen Number of valid samples in `frame`.
 * @param order    LPC order (= 2 * maxFormants).
 * @param coeffs   Output: LPC prediction coefficients, length = order. Overwritten.
 * @param b1       Scratch: forward prediction errors, length ≥ frameLen.
 * @param b2       Scratch: backward prediction errors, length ≥ frameLen.
 * @param aa       Scratch: previous Levinson coefficients, length ≥ order.
 * @returns        Residual mean-square power. 0 if ill-conditioned or silent.
 */
export function burgLpc(
  frame: Float32Array,
  frameLen: number,
  order: number,
  coeffs: Float32Array,
  b1: Float32Array,
  b2: Float32Array,
  aa: Float32Array,
): number {
  for (let j = 0; j < order; j++) coeffs[j] = 0
  if (frameLen <= 2) {
    coeffs[0] = -1.0
    return frameLen === 2
      ? 0.5 * (frame[0]! ** 2 + frame[1]! ** 2)
      : frame[0]! ** 2
  }

  // (3) Total power
  let p = 0
  for (let j = 0; j < frameLen; j++) p += frame[j]! * frame[j]!
  let xms = p / frameLen
  if (xms <= 0) return 0

  // (9) Initialise forward/backward prediction error vectors:
  //   b1[j] = x[j],   b2[j] = x[j+1]   for j = 0..frameLen-2
  b1[0] = frame[0]!
  for (let j = 1; j < frameLen - 1; j++) {
    b1[j] = frame[j]!
    b2[j - 1] = frame[j]!
  }
  b2[frameLen - 2] = frame[frameLen - 1]!

  for (let ii = 0; ii < order; ii++) {
    // (7) Reflection coefficient via Burg's estimate
    const active = frameLen - ii - 1
    let num = 0,
      denom = 0
    for (let j = 0; j < active; j++) {
      num += b1[j]! * b2[j]!
      denom += b1[j]! * b1[j]! + b2[j]! * b2[j]!
    }
    if (denom <= 0) return 0
    coeffs[ii] = (2.0 * num) / denom

    // (10) Update residual power
    xms *= 1.0 - coeffs[ii]! * coeffs[ii]!

    // (5) Levinson recursion: update lower-order coefficients
    //   a[j] = aa[j] − a[ii] * aa[ii−1−j]  for j = 0 .. ii−1
    for (let j = 0; j < ii; j++)
      coeffs[j] = aa[j]! - coeffs[ii]! * aa[ii - 1 - j]!

    if (ii < order - 1) {
      // (8) Save current coefficients; update error vectors.
      //   Note: coeffs[ii] (the reflection coeff) was not touched by step (5).
      for (let j = 0; j <= ii; j++) aa[j] = coeffs[j]!
      const rc = aa[ii]! // reflection coefficient
      const bound = active - 1
      for (let j = 0; j < bound; j++) {
        b1[j]! -= rc * b2[j]!
        b2[j] = b2[j + 1]! - rc * b1[j + 1]! // b1[j+1] is the old value — correct since j+1 > j
      }
    }
  }
  return xms
}

/**
 * Evaluates the LPC analysis polynomial and its derivative at complex z using Horner's method.
 *
 * Polynomial: P(z) = z^n − c[0]·z^(n−1) − c[1]·z^(n−2) − ⋯ − c[n−1]
 * where c = `lpcCoeffs` (output of burgLpc).
 *
 * Writes [P.re, P.im, P'.re, P'.im] into `out[0..3]`.
 */
function evalLpcPoly(
  zRe: number,
  zIm: number,
  lpcCoeffs: Float32Array,
  order: number,
  out: Float32Array,
): void {
  let pRe = 1.0,
    pIm = 0.0,
    dpRe = 0.0,
    dpIm = 0.0
  for (let k = 0; k < order; k++) {
    // P' = P'·z + P  (before updating P)
    const ndpRe = dpRe * zRe - dpIm * zIm + pRe
    const ndpIm = dpRe * zIm + dpIm * zRe + pIm
    dpRe = ndpRe
    dpIm = ndpIm
    // P = P·z − c[k]
    const npRe = pRe * zRe - pIm * zIm - lpcCoeffs[k]!
    const npIm = pRe * zIm + pIm * zRe
    pRe = npRe
    pIm = npIm
  }
  out[0] = pRe
  out[1] = pIm
  out[2] = dpRe
  out[3] = dpIm
}

/**
 * Finds all roots of the LPC analysis polynomial using the Durand-Kerner (Weierstrass)
 * method, then polishes each root with Newton-Raphson.
 *
 * Source (polish): Roots.cpp Roots_Polynomial_polish().
 *
 * @param lpcCoeffs  LPC coefficients from burgLpc(), length = order.
 * @param order      Polynomial degree (LPC order = 2 * maxFormants).
 * @param rootsRe    Output: root real parts, length ≥ order. Overwritten.
 * @param rootsIm    Output: root imaginary parts, length ≥ order. Overwritten.
 * @param peval      Scratch: [pRe, pIm, dpRe, dpIm], length ≥ 4.
 * @returns          true if Durand-Kerner converged within iteration limit.
 */
export function findLpcRoots(
  lpcCoeffs: Float32Array,
  order: number,
  rootsRe: Float32Array,
  rootsIm: Float32Array,
  peval: Float32Array,
): boolean {
  // Initialise roots evenly spaced on a circle inside the unit disk
  for (let k = 0; k < order; k++) {
    const angle = (2 * Math.PI * (k + 0.5)) / order
    rootsRe[k] = 0.9 * Math.cos(angle)
    rootsIm[k] = 0.9 * Math.sin(angle)
  }

  // Durand-Kerner (Gauss-Seidel: update each root immediately)
  let converged = false
  for (let iter = 0; iter < 80 && !converged; iter++) {
    let maxStep2 = 0
    for (let k = 0; k < order; k++) {
      evalLpcPoly(rootsRe[k]!, rootsIm[k]!, lpcCoeffs, order, peval)
      // Denominator = ∏_{j≠k} (z[k] − z[j])
      let denRe = 1.0,
        denIm = 0.0
      for (let j = 0; j < order; j++) {
        if (j === k) continue
        const dRe = rootsRe[k]! - rootsRe[j]!,
          dIm = rootsIm[k]! - rootsIm[j]!
        const nRe = denRe * dRe - denIm * dIm
        denIm = denRe * dIm + denIm * dRe
        denRe = nRe
      }
      const d2 = denRe * denRe + denIm * denIm
      if (d2 === 0) continue
      const stepRe = (peval[0]! * denRe + peval[1]! * denIm) / d2
      const stepIm = (peval[1]! * denRe - peval[0]! * denIm) / d2
      rootsRe[k]! -= stepRe
      rootsIm[k]! -= stepIm
      const s2 = stepRe * stepRe + stepIm * stepIm
      if (s2 > maxStep2) maxStep2 = s2
    }
    if (maxStep2 < 1e-20) converged = true
  }

  // Newton-Raphson polish (up to 80 iterations per root, matching Praat)
  for (let k = 0; k < order; k++) {
    let bestRe = rootsRe[k]!,
      bestIm = rootsIm[k]!
    evalLpcPoly(bestRe, bestIm, lpcCoeffs, order, peval)
    let ymin = Math.hypot(peval[0]!, peval[1]!)

    for (let iter = 0; iter < 80; iter++) {
      evalLpcPoly(rootsRe[k]!, rootsIm[k]!, lpcCoeffs, order, peval)
      const fabsy = Math.hypot(peval[0]!, peval[1]!)
      if (fabsy >= ymin) {
        rootsRe[k] = bestRe
        rootsIm[k] = bestIm
        break
      }
      ymin = fabsy
      bestRe = rootsRe[k]!
      bestIm = rootsIm[k]!
      const dp2 = peval[2]! * peval[2]! + peval[3]! * peval[3]!
      if (dp2 === 0) break
      rootsRe[k]! -= (peval[0]! * peval[2]! + peval[1]! * peval[3]!) / dp2
      rootsIm[k]! -= (peval[1]! * peval[2]! - peval[0]! * peval[3]!) / dp2
    }
  }
  return converged
}

/**
 * Projects roots outside the unit circle to their inverse: z → 1/conj(z).
 * Source: Roots.cpp Roots_fixIntoUnitCircle()
 */
export function fixIntoUnitCircle(
  rootsRe: Float32Array,
  rootsIm: Float32Array,
  n: number,
): void {
  for (let i = 0; i < n; i++) {
    const n2 = rootsRe[i]! * rootsRe[i]! + rootsIm[i]! * rootsIm[i]!
    if (n2 > 1.0) {
      rootsRe[i]! /= n2
      rootsIm[i] = -rootsIm[i]! / n2
    }
  }
}
